catroid/src/androidTest/java/org/catrobat/catroid/retrofittesting/CatroidWebServerAuthenticationTest.kt

```diff
@@ -0,0 +1,189 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.retrofittesting
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.preference.PreferenceManager
+import androidx.test.platform.app.InstrumentationRegistry
+import com.squareup.moshi.Moshi
+import junit.framework.TestCase.assertEquals
+import junit.framework.TestCase.assertNotNull
+import org.catrobat.catroid.common.Constants
+import org.catrobat.catroid.koin.testModules
+import org.catrobat.catroid.retrofit.WebService
+import org.catrobat.catroid.retrofit.models.LoginUser
+import org.catrobat.catroid.retrofit.models.RegisterFailedResponse
+import org.catrobat.catroid.retrofit.models.RegisterUser
+import org.catrobat.catroid.testsuites.annotations.Cat.OutgoingNetworkTests
+import org.catrobat.catroid.web.ServerAuthenticationConstants.SERVER_RESPONSE_REGISTER_OK
+import org.catrobat.catroid.web.ServerAuthenticationConstants.SERVER_RESPONSE_REGISTER_UNPROCESSABLE_ENTITY
+import org.catrobat.catroid.web.ServerAuthenticationConstants.SERVER_RESPONSE_TOKEN_OK
+import org.junit.Before
+import org.junit.Test
+import org.junit.experimental.categories.Category
+import org.junit.runner.RunWith
+import org.koin.core.context.startKoin
+import org.koin.core.context.stopKoin
+import org.koin.test.KoinTest
+import org.koin.test.inject
+import org.mockito.MockitoAnnotations
+import org.mockito.junit.MockitoJUnitRunner
+
+@RunWith(MockitoJUnitRunner::class)
+@Category(OutgoingNetworkTests::class)
+class CatroidWebServerAuthenticationTest : KoinTest {
+
+    companion object {
+        private const val STATUS_CODE_INVALID_CREDENTIALS = 401
+    }
+
+    private lateinit var newEmail: String
+    private lateinit var newUserName: String
+    private var email = "catroweb@localhost.at"
+    private var username = "catroweb"
+    private var invalidUsername = "InvalidUser"
+    private var password = "catroweb"
+    private var wrongPassword = "WrongPassword"
+    private lateinit var context: Context
+    private lateinit var sharedPreferences: SharedPreferences
+
+    private val webServer: WebService by inject()
+
+    @Before
+    fun setUp() {
+        stopKoin()
+        startKoin { modules(testModules) }
+
+        newUserName = "APIUser" + System.currentTimeMillis()
```

Comment:
Should we not use the normal testusers, as discussed, because the api is now available on test api too?

Reply:
No, as this username is used to **register new users** that do not already exist on the server. If I would use a username from an already registered testuser I would get an error (username already exists). At the end of the `testRegistrationOk` the user is deleted anyways so nothing will remain on the server.